zerver/actions/realm_settings.py

```diff
@@ -151,13 +151,17 @@ def do_set_push_notifications_enabled_end_timestamp(
 
 @transaction.atomic(savepoint=False)
 def do_change_realm_permission_group_setting(
-    realm: Realm, setting_name: str, user_group: UserGroup, *, acting_user: Optional[UserProfile]
+    realm: Realm,
+    setting_name: str,
+    user_group: UserGroup,
+    *,
+    old_setting_api_value: int,
+    acting_user: Optional[UserProfile],
```

Comment:
Extracted this change in a separate commit to avoid having tests changes like in `test_invite` in the main commit.

Reply:
Sounds good. It feels a bit inelegant for some of the test use cases to require the caller to pass this in; I wonder if it makes sense to make that parameter optional, and if not provided, this function can just look it up?

Not sure if that's coherent in practice, but definitely something to think about.

Reply:
Updated the old_setting_api_value field to be optional. It is actually fine to compute it in this function itself except for the case when both the old settings and new settings are anonymous groups, because we need to make sure we compute it before the membership of UserGroup value is changed.

So, we just pass it in the actual code always without specifically checking if both old and new values are anonymous groups, and do not pass it in tests, except for one test where we are testing the audit log data when both old and new values are anonymous groups.

Reply:
Yeah, sounds good.

Reply:
Added a bit of comment documentation for this detail:

``` diff
diff --git a/zerver/actions/realm_settings.py b/zerver/actions/realm_settings.py
index 5251e7b84d..0edd0c71b2 100644
--- a/zerver/actions/realm_settings.py
+++ b/zerver/actions/realm_settings.py
@@ -173,6 +173,10 @@ def do_change_realm_permission_group_setting(
     realm.save(update_fields=[setting_name])
 
     if old_setting_api_value is None:
+        # Most production callers will have computed this as part of
+        # verifying whether there's an actual change to make, but it
+        # feels quite clumsy to have to pass it from unit tests, so we
+        # compute it here if not provided by the caller.
         old_setting_api_value = get_group_setting_value_for_api(old_value)
     new_setting_api_value = get_group_setting_value_for_api(user_group)
 
```